front/lib/actions/mcp_internal_actions/servers/webtools_edge.ts

```diff
@@ -0,0 +1,335 @@
+import { DustAPI, INTERNAL_MIME_TYPES } from "@dust-tt/client";
+import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import type { CallToolResult } from "@modelcontextprotocol/sdk/types.js";
+import { z } from "zod";
+
+import {
+  generatePlainTextFile,
+  uploadFileToConversationDataSource,
+} from "@app/lib/actions/action_file_helpers";
+import { MAXED_OUTPUT_FILE_SNIPPET_LENGTH } from "@app/lib/actions/action_output_limits";
+import {
+  WEBBROWSER_TOOL_NAME,
+  WEBSEARCH_TOOL_NAME,
+} from "@app/lib/actions/mcp_internal_actions/constants";
+import {
+  AGENT_CONFIGURATION_URI_PATTERN,
+  ConfigurableToolInputSchemas,
+} from "@app/lib/actions/mcp_internal_actions/input_schemas";
+import type { WebsearchResultResourceType } from "@app/lib/actions/mcp_internal_actions/output_schemas";
+import {
+  makeInternalMCPServer,
+  makeMCPToolTextError,
+} from "@app/lib/actions/mcp_internal_actions/utils";
+import type {
+  AgentLoopContextType,
+  AgentLoopRunContextType,
+} from "@app/lib/actions/types";
+import { getRefs } from "@app/lib/api/assistant/citations";
+import config from "@app/lib/api/config";
+import type { Authenticator } from "@app/lib/auth";
+import { prodAPICredentialsForOwner } from "@app/lib/auth";
+import { concurrentExecutor } from "@app/lib/utils/async_utils";
+import {
+  browseUrls,
+  isBrowseScrapeSuccessResponse,
+} from "@app/lib/utils/webbrowse";
+import { webSearch } from "@app/lib/utils/websearch";
+import logger from "@app/logger/logger";
+import type { Result } from "@app/types";
+import { Err, getHeaderFromUserEmail, Ok } from "@app/types";
+
+// No token-based truncation needed since we don't inline page content in tool output.
+
+function parseAgentConfigurationUri(uri: string): string | null {
+  const match = uri.match(AGENT_CONFIGURATION_URI_PATTERN);
+  return match ? match[2] : null;
+}
+
+async function summarizeWithAgent({
+  auth,
+  agentLoopRunContext,
+  summaryAgentId,
+  content,
+}: {
+  auth: Authenticator;
+  agentLoopRunContext: AgentLoopRunContextType;
+  summaryAgentId: string;
+  content: string;
+}): Promise<Result<string, Error>> {
+  const owner = auth.getNonNullableWorkspace();
+  const user = auth.user();
+  const prodCredentials = await prodAPICredentialsForOwner(owner);
+  const api = new DustAPI(
+    config.getDustAPIConfig(),
+    {
+      ...prodCredentials,
+      extraHeaders: { ...getHeaderFromUserEmail(user?.email) },
+    },
+    logger
+  );
+
+  const mainAgent = agentLoopRunContext.agentConfiguration;
+  const mainConversation = agentLoopRunContext.conversation;
+  const maxChars = 100_000;
+  const toSummarize = content.slice(0, maxChars);
+
+  const convRes = await api.createConversation({
```

Comment:
We probably need to refactor a bit the run_agent code to make it easier to reuse here and in the deep research tool (unmerged yet) -> I can take care of it, WDYT?